crates/shared/src/price_estimation/competition.rs

```diff
@@ -34,62 +28,53 @@ impl RacingCompetitionPriceEstimator {
             successful_results_for_early_return,
         }
     }
+}
 
-    fn enough_successful_estimates_for_each_query(
-        &self,
-        queries: &[Query],
-        results: &[(&String, Vec<PriceEstimateResult>)],
-    ) -> bool {
-        for i in 0..queries.len() {
-            if results.iter().filter(|result| result.1[i].is_ok()).count()
-                < self.successful_results_for_early_return.into()
-            {
-                return false;
-            }
-        }
-
-        true
-    }
-
-    async fn estimates_<'a>(
+impl PriceEstimating for RacingCompetitionPriceEstimator {
+    fn estimates<'a>(
         &'a self,
         queries: &'a [Query],
-    ) -> impl Iterator<Item = PriceEstimateResult> + 'a {
+    ) -> futures::stream::BoxStream<'_, (usize, PriceEstimateResult)> {
         debug_assert!(queries.iter().all(|query| {
             query.buy_token != model::order::BUY_ETH_ADDRESS
                 && query.sell_token != model::order::BUY_ETH_ADDRESS
                 && query.sell_token != query.buy_token
         }));
 
-        let mut remaining_estimates: Vec<_> = self
-            .inner
-            .iter()
-            .map(|(name, estimator)| {
-                async move { (name, vec_estimates(estimator.as_ref(), queries).await) }.boxed()
-            })
-            .collect();
-
-        let mut computed_estimates = Vec::with_capacity(queries.len());
-
-        while !remaining_estimates.is_empty() {
-            let (estimate, _, rest) = future::select_all(remaining_estimates).await;
-            computed_estimates.push(estimate);
-            if self.enough_successful_estimates_for_each_query(queries, &computed_estimates) {
-                break;
+        let combined_stream = futures::stream::select_all(self.inner.iter().enumerate().map(
+            |(i, (_, estimator))| estimator.estimates(queries).map(move |result| (i, result)),
+        ));
+        // Stores the estimates for each query and estimator. When we have collected enough results
+        // to produce a result of our own the corresponding element is set to None.
+        let mut estimates: Vec<Option<Vec<(usize, PriceEstimateResult)>>> =
+            vec![Some(Vec::with_capacity(self.inner.len())); queries.len()];
+        let mut handle_single_result = move |estimator_index: usize, query_index: usize, result| {
```

Comment:
nit: I find `handle_single_result` very hard to parse because everything can happen out of order so we have to juggle 2 indices around.
I think adding an empty line between the different blocks of the function and adding a small comment for each one instead of a single comment for the whole function could be more helpful for people unfamiliar with the code.
Maybe:
1) // Store result and determine if we have enough for this query.
2) // Determine the best result for this query.
3) // Log result.

Reply:
Good suggestion. Will add some comments to make it easier to understand.

Reply:
Thanks. The new comments/white space help me very much. 👍